compiler/src/iree/compiler/Codegen/LLVMCPU/test/aarch64_vector_lowering.mlir

```diff
@@ -171,3 +172,43 @@ module {
+
+func.func @simpul_mul_mixed_mini_dispatch_3_mmt4d_1x1x8x8x8x1_i8xi8xi32() {
+  %c0_i32 = arith.constant 0 : i32
+  %c0_i8 = arith.constant 0 : i8
+  %c8 = arith.constant 8 : index
+  %c1 = arith.constant 1 : index
+  %c0 = arith.constant 0 : index
+  %c64 = arith.constant 64 : index
+  %c128 = arith.constant 128 : index
+  %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : memref<1x8x8x1xi8, #hal.descriptor_type<storage_buffer>>
+  memref.assume_alignment %0, 64 : memref<1x8x8x1xi8, #hal.descriptor_type<storage_buffer>>
+  %1 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c64) flags(ReadOnly) : memref<1x8x8x1xi8, strided<[64, 8, 1, 1], offset: 64>, #hal.descriptor_type<storage_buffer>>
+  memref.assume_alignment %1, 64 : memref<1x8x8x1xi8, strided<[64, 8, 1, 1], offset: 64>, #hal.descriptor_type<storage_buffer>>
+  %2 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c128) : memref<1x1x8x8xi32, strided<[64, 64, 8, 1], offset: 32>, #hal.descriptor_type<storage_buffer>>
+  memref.assume_alignment %2, 64 : memref<1x1x8x8xi32, strided<[64, 64, 8, 1], offset: 32>, #hal.descriptor_type<storage_buffer>>
+  %workgroup_id_x = hal.interface.workgroup.id[0] : index
+  %workgroup_count_x = hal.interface.workgroup.count[0] : index
+  %workgroup_id_y = hal.interface.workgroup.id[1] : index
+  %workgroup_count_y = hal.interface.workgroup.count[1] : index
+  scf.for %arg0 = %workgroup_id_y to %c1 step %workgroup_count_y {
+    scf.for %arg1 = %workgroup_id_x to %c1 step %workgroup_count_x {
+      %3 = vector.transfer_read %2[%c0, %c0, %c0, %c0], %c0_i32 {in_bounds = [true, true, true, true]} : memref<1x1x8x8xi32, strided<[64, 64, 8, 1], offset: 32>, #hal.descriptor_type<storage_buffer>>, vector<1x1x8x8xi32>
+      %4 = scf.for %arg2 = %c0 to %c8 step %c1 iter_args(%arg3 = %3) -> (vector<1x1x8x8xi32>) {
+        %5 = vector.transfer_read %0[%c0, %arg2, %c0, %c0], %c0_i8 {in_bounds = [true, true, true, true]} : memref<1x8x8x1xi8, #hal.descriptor_type<storage_buffer>>, vector<1x1x8x1xi8>
+        %6 = vector.transfer_read %1[%c0, %arg2, %c0, %c0], %c0_i8 {in_bounds = [true, true, true, true]} : memref<1x8x8x1xi8, strided<[64, 8, 1, 1], offset: 64>, #hal.descriptor_type<storage_buffer>>, vector<1x1x8x1xi8>
+        %7 = arith.extsi %5 : vector<1x1x8x1xi8> to vector<1x1x8x1xi32>
+        %8 = arith.extsi %6 : vector<1x1x8x1xi8> to vector<1x1x8x1xi32>
+        %9 = vector.contract {indexing_maps = [affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d2, d3, d5)>, affine_map<(d0, d1, d2, d3, d4, d5) -> (d1, d2, d4, d5)>, affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d3, d4)>], iterator_types = ["parallel", "parallel", "reduction", "parallel", "parallel", "reduction"], kind = #vector.kind<add>} %7, %8, %arg3 : vector<1x1x8x1xi32>, vector<1x1x8x1xi32> into vector<1x1x8x8xi32>
```

Comment:
Do we need a new test? Is adding `CHECK-KERNEL-NOT: llvm.inline_asm asm_dialect` to other cases enough?

Reply:
Unless I am missing something, this test is incorrect. If `--vector-contract-custom-kernels=false` is not used, then we should see e.g. 
```mlir
llvm.inline_asm asm_dialect = att "\0A      fmla $0.4s, $18.4s, $16.s[0]\0A
``` 
instead of `vector.contract`, right? IIUC, that's not happening because:
* inside of [LVMCPUMmt4dVectorLoweringPass::runOnOperation](https://github.com/openxla/iree/blob/6ef1cfe4fe47dadacca62b5bfc96ac5a3a9c4dbf/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUMmt4dVectorLowering.cpp#L97) IREE does run `populateVectorContractCustomKernelsPatterns` (which is meant to add the relevant patterns), but ...
* [populateVectorContractCustomKernelsPatterns](https://github.com/openxla/iree/blob/6ef1cfe4fe47dadacca62b5bfc96ac5a3a9c4dbf/compiler/src/iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp#L1144-L1172) doesn't know what the actual target is and hence adds no patterns.

So, this test is not checking anything specific to this PR. The behaviour tested here can be observed even without this PR.

Reply:
Yeah... the test is incorrect. Maybe the patterns do not match the vector.contract case, like I mentioned -- this is an old path to ukernels.. So I think to properly test it, we can add the `--iree-llvmcpu-enable-vector-contract-custom-kernels` option and add the checks in other lit test cases.

Reply:
Thanks for catching this. I've fixed the tests and added both a positive and negative case to ensure the flag does what it should in both directions.